File: String_to_Integer.java
class Solution {
    public int myAtoi(String str) {
        
        str = str.trim();
        int retval =0;
        if(str.length()==0 || (!Character.isDigit(str.charAt(0))  && str.charAt(0)!='+' && str.charAt(0)!='-' )){
            return retval;
        }
        char sign = str.charAt(0) == '-'? '-':'+';
        int maxLimit = Integer.MAX_VALUE/10;
        
        for(int i=0; i<str.length(); i++){
            char ch= str.charAt(i);
            if(Character.isDigit(ch)){
                if(retval > maxLimit){  // max_value = 12345, 1234 = max limit and retval is 1235  
                    return (sign == '-')?  Integer.MIN_VALUE:Integer.MAX_VALUE   ;
                }
                else if(retval == maxLimit){
                    if(sign == '+'){
                        if(ch >= '7') return Integer.MAX_VALUE;
                    }
                    else {
                        if(ch >= '8') return Integer.MIN_VALUE;
                }
            }
            
            retval = retval*10+ (ch - 48);
            
        } else if (i!= 0) break;
        
     }
        if(sign == '-'){
            return -retval;
        }
        
        return retval;
        
        
    }
}

File: StringToInteger.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Strings
{
    public class StringToInteger
    {
        /*
         * T.C: O(N) where n is chars in string
         * S.C: O(1) 
         */
        public int MyAtoi(string s)
        {
            s = s.Trim();

            if (s == null || s.Length == 0) return 0;

            char sign = '+';

            if (s[0] != '+' && s[0] != '-' && !Char.IsNumber(s[0]))
            {

                return 0;
            }

            int limit = 214748364;
            int result = 0;

            if (s[0] == '-')
            {
                sign = '-';
            }

            for (int i = 0; i < s.Length; i++)
            {
                if (Char.IsNumber(s[i]))
                {
                    if (sign == '-')
                    {
                        if (result > limit)
                        {
                            return int.MinValue;
                        }
                        else if (result == limit)
                        {
                            if ((s[i] - '0') > 8)
                                return int.MinValue;
                        }

                    }
                    else
                    {
                        if (result > limit)
                        {
                            return int.MaxValue;
                        }
                        else if (result == limit)
                        {
                            if ((s[i] - '0') > 7)
                                return int.MaxValue;
                        }
                    }

                    result = result * 10 + (s[i] - '0');

                }
                else if (i != 0 && !char.IsNumber(s[i]))
                {
                    break;
                }

            }

            if (sign == '-')
            {
                return result = result * -1;
            }

            return result;

        }
    }
}
